server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceMetrics.java

```diff
@@ -20,6 +20,11 @@
 import java.util.Map;
 import java.util.concurrent.atomic.AtomicReference;
 
+/**
+ * Maintains balancer metrics and makes them accessible to the {@link MeterRegistry}. Metrics are updated or cleared
+ * (see {@link #zeroAllMetrics()}) as a result of cluster events and the metrics will be pulled for reporting via the MeterRegistry
+ * implementation.
+ */
```

Comment:
The point of this class is that the desired balance metrics are only published by the current master. When the node becomes master it starts returning metrics, but when it loses the leadership it clears its state and stops returning metrics.

The reason is because otherwise the non-master nodes would publish junk metrics (because `updateMetrics` stops being called on non-master nodes), and it makes them nearly impossible to plot.

Reply:
It'd be good to be more explicit about that reasoning/intention in the javadoc

Reply:
Yeah if we return an empty list it means no metrics are published from this node on that poll. So essentially we stop publishing metrics until we're the leader again.

We zero the metrics as well just so that if we do become master again we don't return the stale numbers from the last time we were master. This was just to be consistent-ish with the old behaviour (we used to just return zeros whenever we weren't master). We should probably not publish any metrics until `updateMetrics` is called again really.

There does appear to be a race condition there, but as you say it's not hugely consequential.

Reply:
Updated the text, thanks for explaining!